storage-node/packages/runtime-api/index.js

```diff
@@ -171,119 +170,58 @@ class RuntimeApi {
-          } else if (sudoAsDone) {
-            const dispatchSuccess = sudoAsDone.event.data[0]
-            if (dispatchSuccess.isTrue) {
-              onFinalizedSuccess({ mappedEvents, result, block: status.asFinalized })
-            } else {
-              onFinalizedFailed({ err: 'SudoAsFailed', mappedEvents, result, block: status.asFinalized })
-            }
-          } else {
-            onFinalizedSuccess({ mappedEvents, result, block: status.asFinalized })
-          }
-        }
-      }
+    // object used to communicate back information from the tx updates handler
+    const out = {
+      lastTxUpdateResult: undefined,
     }
 
     // synchronize access to nonce
     await this.executeWithAccountLock(accountId, async () => {
       const nonce = await this.selectBestNonce(accountId)
+      const signed = tx.sign(fromKey, { nonce })
+      const txhash = signed.hash
 
       try {
-        unsubscribe = await tx.sign(fromKey, { nonce }).send(handleTxUpdates)
-        debug('TransactionSubmitted')
-        // transaction submitted successfully, increment and save nonce.
-        this.incrementAndSaveNonce(accountId, nonce)
-      } catch (err) {
-        const errstr = err.toString()
-        debug('TransactionRejected:', errstr)
-        // This happens when nonce is already used in finalized transactions, ie. the selected nonce
-        // was less than current account nonce. A few scenarios where this happens (other than incorrect code)
-        // 1. When a past future tx got finalized because we submitted some transactions
-        // using up the nonces upto that point.
-        // 2. Can happen while storage-node is talkig to a joystream-node that is still not fully
-        // synced.
-        // 3. Storage role account holder sent a transaction just ahead of us via another app.
-        if (errstr.indexOf('ExtrinsicStatus:: 1010: Invalid Transaction: Stale') !== -1) {
-          // In case 1 or 3 a quick recovery could work by just incrementing, but since we
-          // cannot detect which case we are in just reset and force re-reading nonce. Even
-          // that may not be sufficient expect after a few more failures..
-          delete this.nonces[accountId]
-        }
+        callbacks.unsubscribe = await signed.send(
+          RuntimeApi.createTxUpdateHandler(callbacks, { nonce, txhash, subscribed }, out)
+        )
+
+        const serialized = JSON.stringify({
+          nonce,
+          txhash,
+          tx: signed.toHex(),
+        })
 
-        // Technically it means a transaction in the mempool with same
-        // nonce and same fees being paid so we cannot replace it, either we didn't correctly
-        // increment the nonce or someone external to this application sent a transaction
-        // with same nonce ahead of us.
-        if (errstr.indexOf('ExtrinsicStatus:: 1014: Priority is too low') !== -1) {
-          delete this.nonces[accountId]
+        if (out.lastResult.status.isFuture) {
```

Comment:
Also I'm not sure if we can assume that `out.lastResult.status` is already defined here. Although it seems to be the case that `handleTxUpdates` is called before we reach that condition, this is my no means obvious and in TypeScript environment that would be considered an error. Perhaps those conditions can be changed to `out.lastResult && out.lastResult.status && out.lastResult.status.isFuture` for clarity?

Reply:
yes I agree its a very poor approach but unfortunately the only way around it. I'm only confident because I tested the order while debugging this whole codebase. I would prefer to not add the check so we can actually catch this case if it is ever incorrect

Reply:
It is guaranteed that the `Ready` status is called before we reach this code path. When the tx is rejected and therefore the tx isn't added to ready queue and this status is not emitted, then `.send()` throws and so we don't reach this code.
I will add this code comment.